table-api/src/main/java/io/deephaven/api/filter/Filter.java

```diff
@@ -15,8 +17,16 @@
  * Represents an evaluate-able filter.
  *
  * @see io.deephaven.api.TableOperations#where(Collection)
+ * @see FilterIsNull
+ * @see FilterIsNotNull
+ * @see FilterComparison
+ * @see FilterNot
+ * @see FilterOr
+ * @see FilterAnd
+ * @see LiteralFilter
+ * @see RawString
  */
-public interface Filter extends Serializable {
+public interface Filter extends Expression, Serializable {
```

Comment:
So, an `Expression` is something that you can evaluate to _some_ value. `Foo + 32`, `Foo * my_method(Bar)`, `Foo > Bar` are expressions. A `Filter` is a specific expression that always evaluates to a _boolean_. So, `Foo > Bar`, `isNull(Foo)`, `Foo == Bar * 32` are filters.

Given that filtering is such an important concept for table operations, it has its own explicit type - ie, the interface is essentially `Table where(Collection<Filter> filters)` instead of `Table where(Collection<Expression> exprs)`.

Reply:
To clarify, I'm perfectly happy with `Filter` being its own type. What makes me nervous is `Filter` being an `Expression` rather than its own standalone type. It seems like you're saying that Filters are really always just expressions that have a boolean type. But it feels to me like some filters go beyond that.

Counterexample:

in SQL I can say

```
select * from Table where foo in (select something from somewhere)
```

but I can't substitute

```
select foo in (select something from somewhere) from Table
```

at least I don't think I can

The above (if you think it's valid) suggest to me that `Filter` is a different category of thing than Expression.

Reply:
So, in Calcite everything is an "expression" (even what we are calling 'filters"), but they preserve more type information along the way (so, they know returns types, whereas DH is looser in this regards) and can error out when things don't match up. 

Regardless, we _need_ a way to represent assigning a boolean to a column: `table.view("Foo=MyValue>42")`. This is through the lens of `Selectable`, which is essentially a tuple of lhs `ColumnName` and rhs `Expression`. IMO, that means we have 3 options:

1. Have filter be an expression: `class Filter extends Expression`
2. Have filter be wrappable as an expression: `class ExpressionFilter extends Expression { Filter filter(); ...}`
3. Re-create the full structure of filter, but as mirrored onto expression: `ExpressionIsNull`, `ExpressionIsNotNull`, `ExpressionCondition`. The case could also be argued that when it comes to expressions, we should just fall back to the less-strongly-typed construct like `ExpressionFunction`, but I think that means we are depriving ourselves of the structure we already have w/ `Filter`.

My preference is 1 > 2 >> 3.

Reply:
I understand that calcite needs that information, but why does this non-sql api need that? Aren't we converting from SQL to QST, and not vice versa?

Or are you saying that, in context, we can't tell if calcite is going to assign an expression (which happens to be a filter) to a column? And, assuming that is true, can we not instead offer a Filter.toExpression() that will transform from a filter to an expression, rather than conflating the two?

Reply:
It's not that we _need_ `Filter extends Expression`, I just think from a logical typing perspective it's the most natural (I see `Filter` as a specific type of `Expression`).

For example:

```java
table.select(Selectable.of(FOO, isNull(BAR)))
```
is how I imagine we'd get benefit out of having filter naturally be an expression. In the case where it's not and we have a wrapper or a type conversion, it would look like:

```java
table.select(Selectable.of(FOO, isNull(BAR).toExpression()))
```
or
```java
table.select(Selectable.of(FOO, ExpressionFilter.of(isNull(BAR))))
```